### DISCUSSION TITLE
Zustand -> Valtio

### DISCUSSION THREAD

**Discussion by nosferatu500** [pmndrs/valtio#190] [General]
Hi guys, thanks for zustand and valtio.

I started building my application based on zustand. It is a very handy and easy-to-use library.

But I am planning to migrate from zustand to valtio for the following reasons:

1) I use zundo + zustand to manage history (also a very useful plugin), but valtio has it out of the box
2) I have 6 different stores in my project and I am very tired because of the control selectors in zustand (too much code if you need to select something from several stores at the same time). If I understood correctly with valtio, there will be much less code.

But before I want to know about a few things:

1) How can I limit the history of valtio? (example: 100 snapshots as limit, after that replace the oldest one).
2) How can I use valtio with persistent storage? (local storage, indexedDB and etc...)

**Comment by dai-shi** [comment]
Would love to hear your experience on transition.

> How can I limit the history of valtio? (example: 100 snapshots as limit, after that replace the oldest one).

`proxyWithHistory` is fully customizable. (I'd say it's a reference implementation.)
```js
const p = proxyWithHistory(0)
const savedSaveHistory = p.saveHistory
p.saveHistory = () => {
  savedSaveHistory()
  if (p.history.snapshots.length > 100) {
    p.history.snapshots.shift()
    --p.history.index
  }
}
```

> How can I use valtio with persistent storage? (local storage, indexedDB and etc...)

valtio is unopinionated how you accomplish it.
the community make wiki pages.
I just added one for persistence. https://github.com/pmndrs/valtio/wiki/How-to-persist-states It's incomplete.
If you have questions, feel free to ask here or in a new discussion.

**Reply by nosferatu500** [reply]
Thanks for the reply and detailed explanation!

It looks really simple and convenient. I think I'll be back with a result or additional questions by the end of this week. 
Thanks again!

**Reply by nosferatu500** [reply]
@dai-shi Good day!

I have completed the migration to valtio. Everything works smoothly! 
Thanks for the quick fix for the "history issue".

Good luck with this and others projects.

PS: And thanks for your blog. Keep writing please. This is really helpful. I just found it and already learned a lot.

**Reply by dai-shi** [reply]
Good to hear that. You are welcome.
